src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -12330,65 +12221,24 @@ private AssignmentExpressionSyntax ParseDictionaryInitializer()
         private InitializerExpressionSyntax ParseComplexElementInitializer()
         {
             var openBrace = this.EatToken(SyntaxKind.OpenBraceToken);
-            var initializers = _pool.AllocateSeparated<ExpressionSyntax>();
 
-            DiagnosticInfo closeBraceError;
-            this.ParseExpressionsForComplexElementInitializer(ref openBrace, initializers, out closeBraceError);
-            var closeBrace = this.EatToken(SyntaxKind.CloseBraceToken);
-            if (closeBraceError != null)
-            {
-                closeBrace = WithAdditionalDiagnostics(closeBrace, closeBraceError);
-            }
+            // Historically, if we have a trailing separator here, we have reported a special error of "expression
+            // expected" instead of "invalid expression term".  To preserve that, we pass `trailingSeparatorError:
+            // ErrorCode.ERR_ExpressionExpected` below.
+            var initializers = this.ParseCommaSeparatedSyntaxList(
+                ref openBrace,
+                SyntaxKind.CloseBraceToken,
+                static @this => @this.IsPossibleExpression(),
+                static @this => @this.ParseExpressionCore(),
+                static (@this, openBrace, list, expectedKind, closeKind) => @this.SkipBadInitializerListTokens(openBrace, list, expectedKind),
+                allowTrailingSeparator: false,
+                trailingSeparatorError: ErrorCode.ERR_ExpressionExpected);
```

Comment:
so this is something i think is worth bringing up in the PR.  A couple of separated list parsing routines behave subtly different from teh rest.  For example, this location reports a specialized error if there is a trailing comma that is unlike every other location that runs into that issue.  I've preserved this to keep error messages identical.  But, personally, i think it's pointless and we should just unify on the exact same error reporting strategy for all list parsing.  specifically, just call into the element-parsing-function and have it report whatever error it normally would in this case.  

Changing this behavior revealed just 7 tests taht change in the entire compiler.  So i think it's fine to actually update things here, but i wanted to run it by compiler team.